engine/execution/ingestion/engine.go

```diff
@@ -930,344 +992,422 @@ func (e *Engine) generateExecutionReceipt(
+		return fmt.Errorf("invalid origin id (%s): %w", id, err)
+	}
 
-		// since the delta is not an extension of the delta queue, then adding it,
-		// if it doesn't exist before, then add it as a delta orphan
-		// if it exists, then still execute it again.
-		// TODO: maybe we could return nil when !added
-		newQueue := queue.NewQueue(executionStateDelta)
-		added := backdata.Add(newQueue)
-		if added {
-			// add the new branch as a new orphan
-			e.tryRequeueOrphans(executionStateDelta, newQueue, backdata)
-		}
+	// TODO: restrict the sender has to be an execution node.
+	// if id.Role != flow.RoleExecution {
+	// 	return fmt.Errorf("invalid role for requesting state synchronization: %v, %s", originID, id.Role)
+	// }
 
-		// since the delta is not an extension of the delta queue
-		// check if the parent state (StateCommitment) exists in the storage, which would
-		// mean the parent has been executed by ourselves already
-		stateCommitment, getStateCommitmentErr := e.execState.StateCommitmentByBlockID(ctx, executionStateDelta.ParentID())
-		if getStateCommitmentErr != nil && !errors.Is(getStateCommitmentErr, storage.ErrNotFound) {
-			log.Fatal().Msgf("unexpected error while accessing storage for sync deltas, shutting down: %v", getStateCommitmentErr)
-		}
+	// validate that from height must be smaller than to height
+	if req.FromHeight >= req.ToHeight {
+		return engine.NewInvalidInputErrorf("invalid state sync request (from: %x, to: %d)",
+			req.FromHeight, req.ToHeight)
+	}
 
-		// if the parent state for the orphan deltas doesn't exist,
-		// it means this delta is really an orphan, so we stop here, waiting for either
-		// the delta for the parent to come or the parent block to be executed.
-		if errors.Is(getStateCommitmentErr, storage.ErrNotFound) {
-			return nil
-		}
+	lastSealed, err := e.state.Sealed().Head()
+	if err != nil {
+		return fmt.Errorf("could not get last sealed: %w", err)
+	}
 
-		// sanity check that if the state for the parent block already exists,
-		// then the delta's start state must be equal to that.
-		if !bytes.Equal(stateCommitment, executionStateDelta.StartState) {
-			return fmt.Errorf("internal inconsistency with delta for block (%v) - state commitment for parent retrieved from DB (%x) different from start state in delta! (%x)",
-				executionStateDelta.ParentID(),
-				stateCommitment,
-				executionStateDelta.StartState)
-		}
+	sealedHeight := lastSealed.Height
 
-		// if the parent state exists for the orphan delta, we could apply this delta.
-		e.syncWg.Add(1)
-		go e.saveDelta(ctx, executionStateDelta)
+	log := e.log.With().
+		Hex("sender", originID[:]).
+		Uint64("sealed", sealedHeight).
+		Uint64("from", req.FromHeight).
+		Uint64("to", req.ToHeight).
+		Logger()
 
+	// ignore requests for unsealed height
+	if req.FromHeight > sealedHeight {
+		log.Info().Msg("receives state sync requests for unsealed height, ignore")
 		return nil
-	})
+	}
+
+	// fromHeight, toHeight must be sealed height
```

Comment:
We could also do a sanity check it `toHeight` <= `fromHeight`

Reply:
Yeah, [did it here already](https://github.com/onflow/flow-go/pull/31/files#diff-9d11a07967b04bd9f51ebd6b01a170d5R1161)